i18n-helpers/src/lib.rs

```diff
@@ -333,18 +335,139 @@ fn is_comment_skip_directive(html: &str) -> bool {
     re.is_match(html.trim())
 }
 
-/// Returns true if the events appear to be a codeblock without translatable text.
-fn is_nontranslatable_codeblock_group(events: &[(usize, Event)]) -> bool {
-    match events {
-        [(_, Event::Start(Tag::CodeBlock(_))), .., (_, Event::End(Tag::CodeBlock(_)))] => {
-            let (codeblock_text, _) = reconstruct_markdown(events, None);
-            // Heuristic to check whether the codeblock nether has a
-            // literal string nor a line comment.  We may actually
-            // want to use a lexer here to make this more robust.
-            !codeblock_text.contains('"') && !codeblock_text.contains("//")
+/// Returns true if the events appear to be a codeblock.
+fn is_codeblock_group(events: &[(usize, Event)]) -> bool {
+    matches!(
+        events,
+        [
+            (_, Event::Start(Tag::CodeBlock(_))),
+            ..,
+            (_, Event::End(Tag::CodeBlock(_)))
+        ]
+    )
+}
+
+/// Creates groups by parsing codeblock.
+fn parse_codeblock<'a>(events: &'a [(usize, Event)]) -> Vec<Group<'a>> {
+    // Language detection from language identifier of codeblock.
+    let ss = SyntaxSet::load_defaults_newlines();
+    let syntax = if let (_, Event::Start(Tag::CodeBlock(CodeBlockKind::Fenced(x)))) = &events[0] {
+        ss.find_syntax_by_token(x.split(',').next().unwrap())
+    } else {
+        None
+    };
+
+    let is_translate_scope = |x| {
+        let scope_string = Scope::new("string").unwrap();
+        let scope_comment = Scope::new("comment").unwrap();
+        scope_string.is_prefix_of(x) | scope_comment.is_prefix_of(x)
+    };
+
+    let Some(syntax) = syntax else {
+        // If there is no language specifier, the whole codeblock should be translated.
+        return vec![Group::Translate(events.into())];
+    };
+
+    let mut ps = ParseState::new(syntax);
+    let mut ret = vec![];
+
+    for (idx, event) in events.iter().enumerate() {
+        match event {
+            (text_line, Event::Text(text)) => {
+                let mut stack = ScopeStack::new();
+
+                let Ok(ops) = ps.parse_line(text, &ss) else {
+                    // If parse is failed, the whole codeblock should be translated.
+                    ret.push(Group::Translate(events[idx..idx + 1].into()));
+                    continue;
+                };
+
+                let mut translate = vec![];
+
+                for (range, scope) in ScopeRangeIterator::new(&ops, text) {
+                    // Calculate line number of the range
+                    let range_line = if range.start == 0 {
+                        *text_line
+                    } else {
+                        text_line + text[0..range.start].lines().count() - 1
+                    };
+
+                    let _ = stack.apply_with_hook(scope, |op, scopes| {
```

Comment:
I misunderstood the usage of `apply_with_hook`.
If `apply_with_hook` is used, the hook will be called by stack push/pop.
Therefore if the op is `Pop(3)`, the hook is called 3 times. This is not my intention.

I found the mistake through fuzz test, and use `apply` instead of it now.

>  Is it because we're trying to compute line number as well?

Yes

Reply:
I think we can resolve this?